File: packages/python/prompt_sanitizer/engines/regex_engine.py
"""
Regex Engine — Layer 1 of prompt-sanitizer.

Detects structured PII (email, phone, SSN, credit cards, IBANs, IPs,
crypto addresses, MAC addresses, URLs, passport numbers, driving licences,
and date-of-birth patterns) using regular expressions with optional
checksum validation (Luhn for credit cards, IBAN mod-97).

All patterns run on every sanitize() call regardless of Mode.
"""
from __future__ import annotations

import re
from dataclasses import dataclass

from ..entities import EntityType
from ..result import DetectedEntity


# ---------------------------------------------------------------------------
# Luhn algorithm — credit card validation
# ---------------------------------------------------------------------------

def _luhn_valid(card: str) -> bool:
    digits = [int(d) for d in card if d.isdigit()]
    if len(digits) < 13:
        return False
    odd = digits[-1::-2]
    even = digits[-2::-2]
    total = sum(odd) + sum(sum(divmod(d * 2, 10)) for d in even)
    return total % 10 == 0


# ---------------------------------------------------------------------------
# IBAN mod-97 validation
# ---------------------------------------------------------------------------

def _iban_valid(iban: str) -> bool:
    raw = iban.replace(" ", "").replace("-", "").upper()
    if len(raw) < 15 or len(raw) > 34:
        return False
    rearranged = raw[4:] + raw[:4]
    numeric = "".join(str(ord(c) - 55) if c.isalpha() else c for c in rearranged)
    try:
        return int(numeric) % 97 == 1
    except ValueError:
        return False


# ---------------------------------------------------------------------------
# Pattern registry
# ---------------------------------------------------------------------------

@dataclass
class _Pattern:
    entity_type: EntityType
    regex: re.Pattern[str]
    confidence: float
    validator: object = None  # optional callable(match_str) -> bool


# fmt: off
_PATTERNS: list[_Pattern] = [
    # ── Email ────────────────────────────────────────────────────────────────
    _Pattern(
        EntityType.EMAIL,
        re.compile(
            r"(?<![a-zA-Z0-9._%+\-])"
            r"[a-zA-Z0-9._%+\-]{1,64}"
            r"@"
            r"[a-zA-Z0-9.\-]{1,253}"
            r"\.[a-zA-Z]{2,}"
            r"(?![a-zA-Z0-9._%+\-@])",
            re.IGNORECASE,
        ),
        confidence=0.99,
    ),

    # ── US phone (many formats) ──────────────────────────────────────────────
    _Pattern(
        EntityType.PHONE,
        re.compile(
            r"(?<!\d)"
            r"(?:\+?1[\s.\-]?)?"
            r"(?:\([2-9]\d{2}\)|[2-9]\d{2})"
            r"[\s.\-]?"
            r"\d{3}"
            r"[\s.\-]?"
            r"\d{4}"
            r"(?!\d)",
        ),
        confidence=0.85,
    ),

    # ── International phone — compact E.164 e.g. +447946123456 ───────────────
    _Pattern(
        EntityType.PHONE,
        re.compile(r"(?<!\d)\+[1-9]\d{6,14}(?!\d)"),
        confidence=0.80,
    ),

    # ── International phone — spaced/dashed e.g. +44 20 7946 0958 ────────────
    _Pattern(
        EntityType.PHONE,
        re.compile(
            r"(?<!\d)"
            r"\+[1-9]\d{0,3}"           # country code
            r"(?:[\s.\-]\d{2,4}){2,4}"  # 2–4 groups of digits separated by space/dash/dot
            r"(?!\d)",
        ),
        confidence=0.78,
    ),

    # ── US SSN ───────────────────────────────────────────────────────────────
    _Pattern(
        EntityType.SSN,
        re.compile(
            r"(?<!\d)"
            r"(?!000|666|9\d{2})\d{3}"
            r"[\s\-]"
            r"(?!00)\d{2}"
            r"[\s\-]"
            r"(?!0000)\d{4}"
            r"(?!\d)",
        ),
        confidence=0.95,
    ),

    # ── Credit / debit card (Luhn-validated) ─────────────────────────────────
    _Pattern(
        EntityType.CREDIT_CARD,
        re.compile(
            r"(?<!\d)"
            r"(?:4[0-9]{3}|5[1-5][0-9]{2}|3[47][0-9]{2}|3(?:0[0-5]|[68][0-9])[0-9]|"
            r"6(?:011|5[0-9]{2})|(?:2131|1800|35\d{3}))"
            r"[\s\-]?"
            r"(?:\d{4}[\s\-]?){2}"
            r"\d{1,4}"
            r"(?!\d)",
        ),
        confidence=0.95,
        validator=lambda m: _luhn_valid(m),
    ),

    # ── IBAN (mod-97 validated) ───────────────────────────────────────────────
    _Pattern(
        EntityType.IBAN,
        re.compile(
            r"\b[A-Z]{2}\d{2}(?:\s?[A-Z0-9]{4}){2,7}\s?[A-Z0-9]{1,4}\b",
            re.IGNORECASE,
        ),
        confidence=0.92,
        validator=lambda m: _iban_valid(m),
    ),

    # ── IPv4 ─────────────────────────────────────────────────────────────────
    _Pattern(
        EntityType.IP_ADDRESS,
        re.compile(
            r"(?<!\d)"
            r"(?:(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\.){3}"
            r"(?:25[0-5]|2[0-4]\d|[01]?\d\d?)"
            r"(?!\d)",
        ),
        confidence=0.90,
    ),

    # ── IPv6 (full and compressed forms) ─────────────────────────────────────
    _Pattern(
        EntityType.IP_ADDRESS,
        re.compile(
            r"(?<![:\w])"
            r"(?:"
            r"(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}"
            r"|(?:[0-9a-fA-F]{1,4}:){1,7}:"
            r"|(?:[0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}"
            r"|(?:[0-9a-fA-F]{1,4}:){1,5}(?::[0-9a-fA-F]{1,4}){1,2}"
            r"|(?:[0-9a-fA-F]{1,4}:){1,4}(?::[0-9a-fA-F]{1,4}){1,3}"
            r"|::(?:[0-9a-fA-F]{1,4}:){0,5}[0-9a-fA-F]{1,4}"
            r"|[0-9a-fA-F]{1,4}::(?:[0-9a-fA-F]{1,4}:){0,4}[0-9a-fA-F]{1,4}"
            r"|::ffff(?::0{1,4})?:(?:(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\.){3}"
            r"(?:25[0-5]|2[0-4]\d|[01]?\d\d?)"
            r")"
            r"(?![:\w])",
            re.IGNORECASE,
        ),
        confidence=0.90,
    ),

    # ── MAC address ──────────────────────────────────────────────────────────
    _Pattern(
        EntityType.MAC_ADDRESS,
        re.compile(
            r"(?<![:\w])"
            r"(?:[0-9a-fA-F]{2}[:\-]){5}[0-9a-fA-F]{2}"
            r"(?![:\w])",
            re.IGNORECASE,
        ),
        confidence=0.90,
    ),

    # ── URL (http/https) ─────────────────────────────────────────────────────
    _Pattern(
        EntityType.URL,
        re.compile(
            r"https?://"
            r"(?:[a-zA-Z0-9\-._~:/?#\[\]@!$&'()*+,;=%]|(?:%[0-9a-fA-F]{2}))+",
            re.IGNORECASE,
        ),
        confidence=0.85,
    ),

    # ── Bitcoin address (P2PKH, P2SH, Bech32) ────────────────────────────────
    _Pattern(
        EntityType.CRYPTO_ADDRESS,
        re.compile(
            r"(?<![a-zA-Z0-9])"
            r"(?:[13][a-km-zA-HJ-NP-Z1-9]{25,34}"
            r"|bc1[qpzry9x8gf2tvdw0s3jn54khce6mua7l]{6,87})"
            r"(?![a-zA-Z0-9])",
        ),
        confidence=0.88,
    ),

    # ── Ethereum address ─────────────────────────────────────────────────────
    _Pattern(
        EntityType.CRYPTO_ADDRESS,
        re.compile(r"(?<![a-fA-F0-9])0x[0-9a-fA-F]{40}(?![0-9a-fA-F])"),
        confidence=0.92,
    ),

    # ── US Passport ──────────────────────────────────────────────────────────
    _Pattern(
        EntityType.PASSPORT,
        re.compile(r"(?<![A-Z0-9])[A-Z]{1,2}\d{7,9}(?![A-Z0-9])"),
        confidence=0.72,
    ),

    # ── UK Passport ──────────────────────────────────────────────────────────
    _Pattern(
        EntityType.PASSPORT,
        re.compile(r"(?<!\d)\d{9}(?!\d)"),
        confidence=0.60,  # low — 9-digit numbers are common
    ),

    # ── US ZIP code (with context boost — basic pattern only) ─────────────────
    _Pattern(
        EntityType.ZIP_CODE,
        re.compile(r"(?<!\d)\d{5}(?:-\d{4})?(?!\d)"),
        confidence=0.55,  # low standalone confidence
    ),

    # ── Date patterns (various formats) ──────────────────────────────────────
    _Pattern(
        EntityType.DATE,
        re.compile(
            r"(?<!\d)"
            r"(?:"
            r"\d{1,2}[/\-\.]\d{1,2}[/\-\.]\d{2,4}"    # DD/MM/YYYY, MM-DD-YY
            r"|\d{4}[/\-\.]\d{1,2}[/\-\.]\d{1,2}"      # YYYY-MM-DD (ISO)
            r"|(?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)[a-z]*\.?\s+"
            r"\d{1,2},?\s+\d{4}"                        # Month DD, YYYY
            r"|\d{1,2}\s+(?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)[a-z]*"
            r"\s+\d{4}"                                 # DD Month YYYY
            r")"
            r"(?!\d)",
            re.IGNORECASE,
        ),
        confidence=0.75,
    ),
]
# fmt: on


# ---------------------------------------------------------------------------
# Engine
# ---------------------------------------------------------------------------

class RegexEngine:
    """
    Runs all regex patterns against the input text and returns detected entities.

    Custom patterns can be added at runtime via ``add_pattern()``.
    """

    def __init__(self) -> None:
        self._patterns: list[_Pattern] = list(_PATTERNS)
        self._counters: dict[EntityType, int] = {}

    def add_pattern(
        self,
        entity_type: EntityType,
        pattern: str,
        confidence: float = 0.80,
        flags: int = re.IGNORECASE,
    ) -> None:
        """Register a custom regex pattern."""
        self._patterns.append(
            _Pattern(entity_type, re.compile(pattern, flags), confidence)
        )

    def detect(self, text: str) -> list[DetectedEntity]:
        """
        Run all patterns against *text* and return a list of DetectedEntity.

        Overlapping matches from different patterns are kept — deduplication
        happens in the Sanitizer which has the full multi-engine view.
        """
        entities: list[DetectedEntity] = []

        for pat in self._patterns:
            for m in pat.regex.finditer(text):
                matched = m.group(0)

                # Run optional validator (e.g. Luhn check)
                if pat.validator is not None and not pat.validator(matched):
                    continue

                entities.append(
                    DetectedEntity(
                        entity_type=pat.entity_type,
                        original=matched,
                        start=m.start(),
                        end=m.end(),
                        confidence=pat.confidence,
                        layer="regex",
                    )
                )

        return entities

File: packages/ruby/lib/prompt_sanitizer/engines/regex_engine.rb
# frozen_string_literal: true

module PromptSanitizer
  module Engines
    # Regex Engine — Layer 1 of prompt-sanitizer.
    #
    # Detects structured PII (email, phone, SSN, credit cards, IBANs, IPs,
    # crypto addresses, MAC addresses, URLs, passport numbers, driving licences,
    # and date patterns) using regular expressions with optional checksum
    # validation (Luhn for credit cards, IBAN mod-97).
    #
    # All patterns run on every sanitize() call regardless of Mode.
    class RegexEngine
      Pattern = Struct.new(:entity_type, :regex, :confidence, :validator, keyword_init: true)

      # ── Validators ──────────────────────────────────────────────────────────

      # Luhn algorithm — validates credit/debit card numbers.
      def self.luhn_valid?(card)
        digits = card.gsub(/\D/, "").chars.map(&:to_i)
        return false if digits.size < 13

        total = digits.reverse.each_with_index.sum do |d, i|
          i.odd? ? [d * 2 - 9, d * 2].min + (d * 2 >= 10 ? 0 : 0) : d
          # Standard Luhn: double every second digit from the right
          if i.odd?
            doubled = d * 2
            doubled > 9 ? doubled - 9 : doubled
          else
            d
          end
        end
        total % 10 == 0
      end

      # IBAN mod-97 validation.
      def self.iban_valid?(iban)
        raw = iban.gsub(/[\s\-]/, "").upcase
        return false unless raw.length.between?(15, 34)

        rearranged = raw[4..] + raw[0..3]
        numeric    = rearranged.chars.map { |c| c =~ /[A-Z]/ ? (c.ord - 55).to_s : c }.join
        numeric.to_i % 97 == 1
      rescue ArgumentError
        false
      end

      # ── Pattern registry ─────────────────────────────────────────────────────

      PATTERNS = [
        # ── Email ──────────────────────────────────────────────────────────────
        Pattern.new(
          entity_type: EntityType::EMAIL,
          regex:       /(?<![a-zA-Z0-9._%+\-])[a-zA-Z0-9._%+\-]{1,64}@[a-zA-Z0-9.\-]{1,253}\.[a-zA-Z]{2,}(?![a-zA-Z0-9._%+\-@])/i,
          confidence:  0.99
        ),

        # ── US phone ───────────────────────────────────────────────────────────
        Pattern.new(
          entity_type: EntityType::PHONE,
          regex:       /(?<!\d)(?:\+?1[\s.\-]?)?(?:\([2-9]\d{2}\)|[2-9]\d{2})[\s.\-]?\d{3}[\s.\-]?\d{4}(?!\d)/,
          confidence:  0.85
        ),

        # ── International phone — compact E.164 e.g. +447946123456 ────────────
        Pattern.new(
          entity_type: EntityType::PHONE,
          regex:       /(?<!\d)\+[1-9]\d{6,14}(?!\d)/,
          confidence:  0.80
        ),

        # ── International phone — spaced/dashed e.g. +44 20 7946 0958 ─────────
        Pattern.new(
          entity_type: EntityType::PHONE,
          regex:       /(?<!\d)\+[1-9]\d{0,3}(?:[\s.\-]\d{2,4}){2,4}(?!\d)/,
          confidence:  0.78
        ),

        # ── US SSN ─────────────────────────────────────────────────────────────
        Pattern.new(
          entity_type: EntityType::SSN,
          regex:       /(?<!\d)(?!000|666|9\d{2})\d{3}[\s\-](?!00)\d{2}[\s\-](?!0000)\d{4}(?!\d)/,
          confidence:  0.95
        ),

        # ── Credit / debit card (Luhn-validated) ──────────────────────────────
        Pattern.new(
          entity_type: EntityType::CREDIT_CARD,
          regex:       /(?<!\d)(?:4[0-9]{3}|5[1-5][0-9]{2}|3[47][0-9]{2}|3(?:0[0-5]|[68][0-9])[0-9]|6(?:011|5[0-9]{2})|(?:2131|1800|35\d{3}))[\s\-]?(?:\d{4}[\s\-]?){2}\d{1,4}(?!\d)/,
          confidence:  0.95,
          validator:   ->(m) { luhn_valid?(m) }
        ),

        # ── IBAN (mod-97 validated) ────────────────────────────────────────────
        Pattern.new(
          entity_type: EntityType::IBAN,
          regex:       /\b[A-Z]{2}\d{2}(?:\s?[A-Z0-9]{4}){2,7}\s?[A-Z0-9]{1,4}\b/i,
          confidence:  0.92,
          validator:   ->(m) { iban_valid?(m) }
        ),

        # ── IPv4 ───────────────────────────────────────────────────────────────
        Pattern.new(
          entity_type: EntityType::IP_ADDRESS,
          regex:       /(?<!\d)(?:(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\.){3}(?:25[0-5]|2[0-4]\d|[01]?\d\d?)(?!\d)/,
          confidence:  0.90
        ),

        # ── IPv6 (full and compressed forms) ──────────────────────────────────
        Pattern.new(
          entity_type: EntityType::IP_ADDRESS,
          regex:       /(?<![:\w])(?:(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}|(?:[0-9a-fA-F]{1,4}:){1,7}:|(?:[0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|(?:[0-9a-fA-F]{1,4}:){1,5}(?::[0-9a-fA-F]{1,4}){1,2}|::(?:[0-9a-fA-F]{1,4}:){0,5}[0-9a-fA-F]{1,4}|[0-9a-fA-F]{1,4}::(?:[0-9a-fA-F]{1,4}:){0,4}[0-9a-fA-F]{1,4})(?![:\w])/i,
          confidence:  0.90
        ),

        # ── MAC address ────────────────────────────────────────────────────────
        Pattern.new(
          entity_type: EntityType::MAC_ADDRESS,
          regex:       /(?<![:\w])(?:[0-9a-fA-F]{2}[:\-]){5}[0-9a-fA-F]{2}(?![:\w])/i,
          confidence:  0.90
        ),

        # ── URL (http/https) ───────────────────────────────────────────────────
        Pattern.new(
          entity_type: EntityType::URL,
          regex:       /https?:\/\/(?:[a-zA-Z0-9\-._~:\/?#\[\]@!$&'()*+,;=%]|(?:%[0-9a-fA-F]{2}))+/i,
          confidence:  0.85
        ),

        # ── Bitcoin address (P2PKH, P2SH, Bech32) ─────────────────────────────
        Pattern.new(
          entity_type: EntityType::CRYPTO_ADDRESS,
          regex:       /(?<![a-zA-Z0-9])(?:[13][a-km-zA-HJ-NP-Z1-9]{25,34}|bc1[qpzry9x8gf2tvdw0s3jn54khce6mua7l]{6,87})(?![a-zA-Z0-9])/,
          confidence:  0.88
        ),

        # ── Ethereum address ───────────────────────────────────────────────────
        Pattern.new(
          entity_type: EntityType::CRYPTO_ADDRESS,
          regex:       /(?<![a-fA-F0-9])0x[0-9a-fA-F]{40}(?![0-9a-fA-F])/,
          confidence:  0.92
        ),

        # ── US Passport ────────────────────────────────────────────────────────
        Pattern.new(
          entity_type: EntityType::PASSPORT,
          regex:       /(?<![A-Z0-9])[A-Z]{1,2}\d{7,9}(?![A-Z0-9])/,
          confidence:  0.72
        ),

        # ── US ZIP code ────────────────────────────────────────────────────────
        Pattern.new(
          entity_type: EntityType::ZIP_CODE,
          regex:       /(?<!\d)\d{5}(?:-\d{4})?(?!\d)/,
          confidence:  0.55
        ),

        # ── Date patterns (DD/MM/YYYY, YYYY-MM-DD, Month DD YYYY, etc.) ────────
        Pattern.new(
          entity_type: EntityType::DATE,
          regex:       /(?<!\d)(?:\d{1,2}[\/\-\.]\d{1,2}[\/\-\.]\d{2,4}|\d{4}[\/\-\.]\d{1,2}[\/\-\.]\d{1,2}|(?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)[a-z]*\.?\s+\d{1,2},?\s+\d{4}|\d{1,2}\s+(?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)[a-z]*\s+\d{4})(?!\d)/i,
          confidence:  0.75
        ),
      ].freeze

      # ── Instance ─────────────────────────────────────────────────────────────

      def initialize
        @patterns = PATTERNS.dup
      end

      # Register a custom regex pattern at runtime.
      #
      #   engine.add_pattern(:custom, /\bACME-\d{6}\b/, confidence: 0.90)
      def add_pattern(entity_type, regex, confidence: 0.80, validator: nil)
        @patterns << Pattern.new(
          entity_type: entity_type,
          regex:       regex,
          confidence:  confidence,
          validator:   validator
        )
      end

      # Run all patterns against +text+ and return an Array of DetectedEntity.
      # Overlapping matches from different patterns are kept — deduplication
      # happens in the Sanitizer, which has the full multi-engine view.
      def detect(text)
        safe_text = text.encode("UTF-8", invalid: :replace, undef: :replace, replace: "")
        entities  = []

        @patterns.each do |pat|
          safe_text.scan(pat.regex) do
            m     = Regexp.last_match
            value = m[0]

            next if pat.validator && !self.class.instance_exec(value, &pat.validator)

            entities << DetectedEntity.new(
              entity_type: pat.entity_type,
              original:    value,
              replacement: nil,
              start_pos:   m.begin(0),
              end_pos:     m.end(0),
              confidence:  pat.confidence,
              layer:       :regex
            )
          end
        end

        entities
      end
    end
  end
end
